Tests.CSharp/Homework3/SingleInitializationSingleton.cs

```diff
@@ -16,15 +17,40 @@ private SingleInitializationSingleton(int delay = DefaultDelay)
 
     public int Delay { get; }
 
-    public static SingleInitializationSingleton Instance => throw new NotImplementedException();
+    public static SingleInitializationSingleton Instance => _instance.Value;
 
     internal static void Reset()
     {
-        throw new NotImplementedException();
+        if (_isInitialized)
+        {
+            lock (Locker)
+            {
+                if (_isInitialized)
+                {
+                    _instance = new Lazy<SingleInitializationSingleton>();
+                    _isInitialized = false;
+                }
+            }
+        }
     }
 
     public static void Initialize(int delay)
     {
-        throw new NotImplementedException();
+        if (!_isInitialized)
+        {
+            lock (Locker)
+            {
+                if (!_isInitialized)
+                {
+                    _instance = new Lazy<SingleInitializationSingleton>(() =>
+                        new SingleInitializationSingleton(delay));
+                    _isInitialized = true;
+                }
+            }
+        }
+        else
+        {
+            throw new InvalidOperationException("Already was initialized");
+        }
```

Comment:
Можно упростить восприятие
```suggestion
        if (_isInitialized)
        {
            throw new InvalidOperationException("Already was initialized");
        }
        
        lock (Locker)
        {
            if (!_isInitialized)
            {
                _instance = new Lazy<SingleInitializationSingleton>(() =>
                    new SingleInitializationSingleton(delay));
                _isInitialized = true;
            }
        }
```